test/unit/org/apache/cassandra/io/util/ThreadLocalReadAheadBufferTest.java

```
@@ -0,0 +1,184 @@
+                    throw new RuntimeException(e);
+                }
+
+                Assert.assertEquals(buf1, buf2);
+            }
+        }
+    }
+
+    private Gen<InputData> lastBlockReads()
+    {
+        return arbitrary().pick(List.of(files))
+                          .flatMap((file) ->
+                                   lists().of(longs().between(0, fileSize(file)).zip(integers().between(1, 100), Pair::create))
+                                          .ofSizeBetween(5, 10)
+                                          .map(positionsAndLengths -> new InputData(file, positionsAndLengths)));
+
+    }
+
+    private Gen<InputData> randomReads()
+    {
+        int blockSize = new DataStorageSpec.IntKibibytesBound("256KiB").toBytes();
+        return arbitrary().pick(List.of(files))
+                         .flatMap((file) ->
+                                  lists().of(longs().between(fileSize(file) - blockSize, fileSize(file)).zip(integers().between(1, 100), Pair::create))
+                                         .ofSizeBetween(5, 10)
+                                         .map(positionsAndLengths -> new InputData(file, positionsAndLengths)));
+
+    }
+
+    // need this becasue generators don't handle the IOException
+    private long fileSize(File file)
+    {
+        try
+        {
+            return Files.size(file.toPath());
+        } catch (IOException e)
+        {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static class InputData
+    {
+
+        private final File file;
+        private final List<Pair<Long, Integer>> positionsAndLengths;
+
+        public InputData(File file, List<Pair<Long, Integer>> positionsAndLengths)
+        {
+            this.file = file;
+            this.positionsAndLengths = positionsAndLengths;
+        }
+    }
+
+    private static File writeFile(int seed, int length)
+    {
+        String fileName = JAVA_IO_TMPDIR.getString() + "data+" + length + ".bin";
```

I were very-very unlucky :-) and got the same random number twice while running the test, in this case the test fails due to repeating file names, it would be better to use a sequence to name the files:
```
at org.apache.cassandra.io.util.ThreadLocalReadAheadBufferTest.cleanup(ThreadLocalReadAheadBufferTest.java:67)
...
Caused by: java.nio.file.NoSuchFileException: /var/folders/d2/pzjmpmzd71j7qp55ry4b98qw0000gn/T/data+2097151.bin
	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:86)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
	at sun.nio.fs.UnixFileSystemProvider.implDelete(UnixFileSystemProvider.java:244)
	at sun.nio.fs.AbstractFileSystemProvider.delete(AbstractFileSystemProvider.java:103)
	at java.nio.file.Files.delete(Files.java:1126)
	at org.apache.cassandra.io.util.PathUtils.delete(PathUtils.java:246)
	... 21 more
```

---

I think we could just catch and ignore this exception. IF we get two files with the same length we just have a bit less entropy in the test because there will be one less file. we just need to account for it happening.

---

See https://github.com/apache/cassandra/pull/3606/commits/38d47aae5d6a2ee36fdf93ba3c467e9902043964

---

yes, I think catching and ignoring the exception is ok. I would suggest also to log seed used by the Random to be able to reproduce other test errors if we get them in future.

---

Added log